KAFKA-18913: Removing _state.updater.enabled_ flag through the Stream…

janchilling opened the pull request:
…Thread#runOnceWithoutProcessingThreads flow.

Removed code of the method StreamThread#initializeAndRestorePhase, since
it was only used in the negation flow. Also removed the flag entirely
from the StreamThread#runOnceWithoutProcessingThreads method.

Will remove the flag and the related code entirely from the future
commits!

janchilling committed:
KAFKA-18913: Removing _state.updater.enabled_ flag through the StreamThread#runOnceWithoutProcessingThreads flow.

Removed code of the method StreamThread#initializeAndRestorePhase, since it was only used in the negation flow. Also removed the flag entirely from the StreamThread#runOnceWithoutProcessingThreads method.

janchilling commented:
Hi @cadonna ,

Would you be able to verify if the changes I have done are correct, like is this what is required?

Also TaskManager#needsInitializationOrRestoration had only one usage through StreamThread#initializeAndRestorePhase method (which was removed). And TaskManager#tryToCompleteRestoration also had only one usage but lot of test cases around 53 I guess. So wanted to know if I should go levels below and remove the code or just top level code.

I will do the changes to remove the flag from the other places and commit them, in the meantime would be great if you can verify if my approach is correct. Also could you also start the workflow as well?

janchilling committed:
KAFKA-18913: Removing _state.updater.enabled_ flag through the StreamThread#create flow.

Removed the stateUpdaterEnabled dependency completely from the StreamThread#create methods and all the inner methods that use it

janchilling committed:
KAFKA-18913: Removing _state.updater.enabled_ flag through the StreamThread#maybeGetClientInstanceIds,  StreamThread#pollPhase and StreamThread#clientInstanceIds flow.

Removed the stateUpdaterEnabled dependency completely from the StreamThread#maybeGetClientInstanceIds,  StreamThread#pollPhase and StreamThread#clientInstanceIds flow.

Also with this commit, have removed the flag dependency completely from the StreamThread and StreamThreadTest classes

janchilling committed:
KAFKA-18913: Removed an unused import from the the StreamThreadTest.

janchilling committed:
KAFKA-18913: CheckStyle updated to correct a failing build

cadonna commented on streams/src/main/java/org/apache/kafka/streams/processor/internals/StreamThread.java:
```diff
@@ -372,15 +371,13 @@ public static StreamThread create(final TopologyMetadata topologyMetadata,
                                       final Runnable shutdownErrorHook,
                                       final BiConsumer<Throwable, Boolean> streamsUncaughtExceptionHandler) {
 
-        final boolean stateUpdaterEnabled = InternalConfig.stateUpdaterEnabled(config.originals());
```
Could you please also remove the internal config `_state.updater.enabled_` and all corresponding code from `StreamsConfig`?

janchilling replied:
Will attend to this within this week

cadonna replied:
There is still a system test that uses the config. It is [`streams_upgrade_test.test_upgrade_downgrade_state_updater()`](https://github.com/apache/kafka/blob/1ded681684e771b16aa98ae751f39b9816345a83/tests/kafkatest/tests/streams/streams_upgrade_test.py#L178). There is a comment that says:
```
Once same-thread state restoration is removed from the code, this test 
should use different versions of the code.
``` 
I guess it means to only use a version before `3.8` (e.g. `LATEST_3_7`) for the `from_version` and `DEV_VERSION` for the `to_version`. You need to choose a version before `3.8` because before `3.8` the state updater was not enabled by default. 
@lucasbru did I correctly interpret your comment?

cadonna commented on streams/src/main/java/org/apache/kafka/streams/processor/internals/StreamThread.java:
```diff
@@ -237,13 +237,13 @@ State setState(final State newState) {
             if (state == State.PENDING_SHUTDOWN && newState != State.DEAD) {
                 log.debug("Ignoring request to transit from PENDING_SHUTDOWN to {}: " +
-                              "only DEAD state is a valid next state", newState);
+                        "only DEAD state is a valid next state", newState);
```
Could you revert this change, please?

cadonna commented on streams/src/main/java/org/apache/kafka/streams/processor/internals/StreamThread.java:
```diff
@@ -237,13 +237,13 @@ State setState(final State newState) {
             } else if (state == State.DEAD) {
                 log.debug("Ignoring request to transit from DEAD to {}: " +
-                              "no valid next state after DEAD", newState);
+                        "no valid next state after DEAD", newState);
```
Could you revert this change, please?

cadonna commented on streams/src/main/java/org/apache/kafka/streams/processor/internals/StreamThread.java:
```diff
@@ -394,76 +391,73 @@ public static StreamThread create(final TopologyMetadata topologyMetadata,
+                adminClient,
+                restoreConsumer,
+                userStateRestoreListener,
+                userStandbyUpdateListener
```
Could you please revert this change?

cadonna commented on streams/src/main/java/org/apache/kafka/streams/processor/internals/StreamThread.java:
```diff
@@ -394,76 +391,73 @@ public static StreamThread create(final TopologyMetadata topologyMetadata,
+                threadIdx,
+                processId,
+                log,
+                proceessingThreadsEnabled
```
Could you please revert the additional indentation?

cadonna commented on streams/src/main/java/org/apache/kafka/streams/processor/internals/StreamThread.java:
```diff
@@ -394,76 +391,73 @@ public static StreamThread create(final TopologyMetadata topologyMetadata,
+                stateDirectory,
+                changelogReader,
+                threadId,
+                log);
```
Could you please revert the additional indentation?

cadonna commented on streams/src/main/java/org/apache/kafka/streams/processor/internals/StreamThread.java:
```diff
@@ -394,76 +391,73 @@ public static StreamThread create(final TopologyMetadata topologyMetadata,
         final Tasks tasks = new Tasks(new LogContext(logPrefix));
         final boolean processingThreadsEnabled =
-            InternalConfig.processingThreadsEnabled(config.originals());
+                InternalConfig.processingThreadsEnabled(config.originals());
```
Could you please revert the additional indentation?

cadonna commented on streams/src/main/java/org/apache/kafka/streams/processor/internals/StreamThread.java:
```diff
@@ -394,76 +391,73 @@ public static StreamThread create(final TopologyMetadata topologyMetadata,
 
         final DefaultTaskManager schedulingTaskManager =
-            maybeCreateSchedulingTaskManager(processingThreadsEnabled, stateUpdaterEnabled, topologyMetadata, time, threadId, tasks);
+                maybeCreateSchedulingTaskManager(processingThreadsEnabled, topologyMetadata, time, threadId, tasks);
```
Could you please revert the additional indentation?

cadonna commented on streams/src/main/java/org/apache/kafka/streams/processor/internals/StreamThread.java:
```diff
@@ -394,76 +391,73 @@ public static StreamThread create(final TopologyMetadata topologyMetadata,
+                        time,
+                        clientId,
+                        threadIdx
+                );
```
Could you please revert the additional indentation?

cadonna commented on streams/src/main/java/org/apache/kafka/streams/processor/internals/StreamThread.java:
```diff
@@ -394,76 +391,73 @@ public static StreamThread create(final TopologyMetadata topologyMetadata,
+                adminClient,
+                stateDirectory,
+                stateUpdater,
+                schedulingTaskManager
```
Could you please revert the additional indentation?

cadonna commented on streams/src/main/java/org/apache/kafka/streams/processor/internals/StreamThread.java:
```diff
@@ -486,81 +480,73 @@ public static StreamThread create(final TopologyMetadata topologyMetadata,
+                referenceContainer.nonFatalExceptionsToHandle,
+                shutdownErrorHook,
+                streamsUncaughtExceptionHandler,
+                cache::resize
```
Could you please revert the additional indentation?

cadonna commented on streams/src/main/java/org/apache/kafka/streams/processor/internals/StreamThread.java:
```diff
@@ -486,81 +480,73 @@ public static StreamThread create(final TopologyMetadata topologyMetadata,
+                    tasks,
+                    new DefaultTaskExecutorCreator(),
+                    topologyMetadata.taskExecutionMetadata(),
+                    1
```
Could you please revert the additional indentation?

cadonna commented on streams/src/main/java/org/apache/kafka/streams/processor/internals/StreamThread.java:
```diff
@@ -486,81 +480,73 @@ public static StreamThread create(final TopologyMetadata topologyMetadata,
 
-    private static StateUpdater maybeCreateAndStartStateUpdater(final boolean stateUpdaterEnabled,
-                                                                final StreamsMetricsImpl streamsMetrics,
+    private static StateUpdater maybeCreateAndStartStateUpdater(final StreamsMetricsImpl streamsMetrics,
```
Since this method now always creates and starts a state updater, could you please rename it to `createAndStartStateUpdater()`?

janchilling replied:
Updated the method name!

cadonna commented on streams/src/main/java/org/apache/kafka/streams/processor/internals/StreamThread.java:
```diff
@@ -615,27 +601,27 @@ public StreamThread(final Time time,
-            time.milliseconds()
+                threadId,
+                streamsMetrics,
+                time.milliseconds()
```
Could you please revert the additional indentation?

cadonna commented on streams/src/main/java/org/apache/kafka/streams/processor/internals/StreamThread.java:
```diff
@@ -615,27 +601,27 @@ public StreamThread(final Time time,
+                processId.toString(),
+                threadId,
+                streamsMetrics,
+                (metricConfig, now) -> this.state().ordinal());
```
Could you please revert the additional indentation?

cadonna commented on streams/src/main/java/org/apache/kafka/streams/processor/internals/StreamThread.java:
```diff
@@ -615,27 +601,27 @@ public StreamThread(final Time time,
-            (metricConfig, now) -> this.state());
+                threadId,
+                streamsMetrics,
+                (metricConfig, now) -> this.state());
```
Could you please revert the additional indentation?

cadonna commented on streams/src/main/java/org/apache/kafka/streams/processor/internals/StreamThread.java:
```diff
@@ -615,27 +601,27 @@ public StreamThread(final Time time,
+                        restoreConsumer,
+                        taskManager::totalProducerBlockedTime
+                ),
+                streamsMetrics
```
Could you please revert the additional indentation?

cadonna commented on streams/src/main/java/org/apache/kafka/streams/processor/internals/StreamThread.java:
```diff
@@ -657,13 +643,12 @@ public StreamThread(final Time time,
         final int dummyThreadIdx = 1;
         this.maxPollTimeMs = new InternalConsumerConfig(config.getMainConsumerConfigs("dummyGroupId", "dummyClientId", dummyThreadIdx))
-            .getInt(ConsumerConfig.MAX_POLL_INTERVAL_MS_CONFIG);
+                .getInt(ConsumerConfig.MAX_POLL_INTERVAL_MS_CONFIG);
```
Could you please revert the additional indentation?

cadonna commented on streams/src/main/java/org/apache/kafka/streams/processor/internals/StreamThread.java:
```diff
@@ -743,7 +728,7 @@ boolean runLoop() {
             } catch (final TaskCorruptedException e) {
                 log.warn("Detected the states of tasks " + e.corruptedTasks() + " are corrupted. " +
-                         "Will close the task as dirty and re-create and bootstrap from scratch.", e);
+                        "Will close the task as dirty and re-create and bootstrap from scratch.", e);
```
Could you please revert the additional indentation?

cadonna commented on streams/src/main/java/org/apache/kafka/streams/processor/internals/StreamThread.java:
```diff
@@ -760,12 +745,12 @@ boolean runLoop() {
                 final String errorMessage = e.getMessage();
                 if (errorMessage != null &&
-                    errorMessage.startsWith("Broker unexpectedly doesn't support requireStable flag on version ")) {
+                        errorMessage.startsWith("Broker unexpectedly doesn't support requireStable flag on version ")) {
```
Could you please revert the additional indentation?

cadonna commented on streams/src/main/java/org/apache/kafka/streams/processor/internals/StreamThread.java:
```diff
@@ -760,12 +745,12 @@ boolean runLoop() {
-                          StreamsConfig.EXACTLY_ONCE_V2);
+                                    "Setting {}=\"{}\" requires broker version 2.5 or higher.",
+                            StreamsConfig.PROCESSING_GUARANTEE_CONFIG,
+                            StreamsConfig.EXACTLY_ONCE_V2);
```
Could you please revert the additional indentation?

cadonna commented on streams/src/main/java/org/apache/kafka/streams/processor/internals/StreamThread.java:
```diff
@@ -801,13 +786,13 @@ void maybeGetClientInstanceIds() {
                 } else {
                     mainConsumerInstanceIdFuture.completeExceptionally(
-                        new TimeoutException("Could not retrieve main consumer client instance id.")
+                            new TimeoutException("Could not retrieve main consumer client instance id.")
```
Could you please revert the additional indentation?

cadonna commented on streams/src/main/java/org/apache/kafka/streams/processor/internals/StreamThread.java:
```diff
@@ -801,13 +786,13 @@ void maybeGetClientInstanceIds() {
 
 
-            if (!stateUpdaterEnabled && !restoreConsumerInstanceIdFuture.isDone()) {
+            if (!restoreConsumerInstanceIdFuture.isDone()) {
```
Shouldn't you delete the whole `if`-clause?
Since the state updater will be always enabled after this PR, this block should always be skipped.

janchilling replied:
Ah yeah makes sense, removed the entire condition and is updated in the latest commit. There was another condition where part of it checked for the same `!stateUpdaterEnabled && !restoreConsumerInstanceIdFuture.isDone()` condition in the latter part of the same method. Remove that condition also since it will always be false.

cadonna commented on streams/src/main/java/org/apache/kafka/streams/processor/internals/StreamThread.java:
```diff
@@ -822,7 +807,7 @@ void maybeGetClientInstanceIds() {
                 } else {
                     restoreConsumerInstanceIdFuture.completeExceptionally(
-                        new TimeoutException("Could not retrieve restore consumer client instance id.")
+                            new TimeoutException("Could not retrieve restore consumer client instance id.")
```
Could you please revert the additional indentation?

cadonna commented on streams/src/main/java/org/apache/kafka/streams/processor/internals/StreamThread.java:
```diff
@@ -831,7 +816,7 @@ void maybeGetClientInstanceIds() {
                     try {
                         producerInstanceIdFuture.complete(
-                            taskManager.streamsProducer().kafkaProducer().clientInstanceId(Duration.ZERO)
+                                taskManager.streamsProducer().kafkaProducer().clientInstanceId(Duration.ZERO)
```
Could you please revert the additional indentation?

cadonna commented on streams/src/main/java/org/apache/kafka/streams/processor/internals/StreamThread.java:
```diff
@@ -844,14 +829,14 @@ void maybeGetClientInstanceIds() {
                 } else {
                     producerInstanceIdFuture.completeExceptionally(
-                        new TimeoutException("Could not retrieve thread producer client instance id.")
+                            new TimeoutException("Could not retrieve thread producer client instance id.")
```
Could you please revert the additional indentation?

cadonna commented on streams/src/main/java/org/apache/kafka/streams/processor/internals/StreamThread.java:
```diff
@@ -844,14 +829,14 @@ void maybeGetClientInstanceIds() {
-                && (!stateUpdaterEnabled && restoreConsumerInstanceIdFuture.isDone())
-                && producerInstanceIdFuture.isDone()) {
+                    && restoreConsumerInstanceIdFuture.isDone()
+                    && producerInstanceIdFuture.isDone()) {
```
Could you please revert the additional indentation?

cadonna commented on streams/src/main/java/org/apache/kafka/streams/processor/internals/StreamThread.java:
```diff
@@ -844,14 +829,14 @@ void maybeGetClientInstanceIds() {
             if (mainConsumerInstanceIdFuture.isDone()
-                && (!stateUpdaterEnabled && restoreConsumerInstanceIdFuture.isDone())
-                && producerInstanceIdFuture.isDone()) {
+                    && restoreConsumerInstanceIdFuture.isDone()
```
I believe you can also remove this line. With the state updater fetching this instance ID is done within the state updater.

cadonna commented on streams/src/main/java/org/apache/kafka/streams/processor/internals/StreamThread.java:
```diff
@@ -918,8 +903,8 @@ public void sendShutdownRequest(final AssignorError assignorError) {
-                     "This implies that this thread missed a rebalance and dropped out of the consumer group. " +
-                     "Will close out all assigned tasks and rejoin the consumer group.", e);
+                "This implies that this thread missed a rebalance and dropped out of the consumer group. " +
+                "Will close out all assigned tasks and rejoin the consumer group.", e);
```
Could you please revert the additional indentation?

cadonna reviewed (commented):
Thanks for the PR, @janchilling !

I started the review, but was not able to finish it. Nevertheless, I would like to share the comments up until now with you. Most comments are about additional indentation. We use 4 spaces and not 8. There are also some other comments about the code.

I will proceed with the review next week. Feel free to update the PR in the meanwhile.

janchilling committed:
KAFKA-18913: Indentation fixes from an earlier commit

janchilling commented:
Hi @cadonna ,

Thank you for the review!

Really sorry that you had to go through the indentation thing, I must have selected the whole file by accident and clicked ctrl+alt+l!

Anyway the indentation thing is now resolved in all places and some other reviews has also been resolved now. I'll attend to the rest(Removing the flag from the rest of the codebase) within this week.

janchilling committed:
KAFKA-18913: Updated the method name of maybeCreateAndStartStateUpdater to createAndStartStateUpdater since it now always creates and starts the state updater

janchilling committed:
KAFKA-18913: With this commit removed a redundant block of code and updated a redundant part of a condition check and an indentation update

janchilling committed:
KAFKA-18913: With this commit removed a redundant block of code.

janchilling committed:
KAFKA-18913: stateuUpdaterEnabled removed from the StateChangelogReader and StateChangeLogReaderTest classes.

Removed the boolean for useNonBlockingPoll since it will always be false(!stateUpdaterEnabled is always false). Therefore, the poll method will always use the polltime. Test cases in the StateChangelogReaderTest has also been updated to match the above scenario. In the Test class the methods shouldPollWithRightTimeoutWithoutStateUpdater and shouldPollWithRightTimeout(final boolean stateUpdaterEnabled, final Task.TaskType type) is redundant since the current default behavior of stateUpdater always true will be checked by shouldPollWithRightTimeoutWithStateUpdaterDefault method.

janchilling committed:
KAFKA-18913: Update for a workflow failure due to an unused import in the StateChangeLogReader and StateChangeLogReaderTest classes

cadonna commented on streams/src/main/java/org/apache/kafka/streams/processor/internals/ProcessorStateManager.java:
```diff
@@ -616,7 +598,7 @@ public void flushCache() {
-        if (!stateUpdaterEnabled && changelogReader != null) {
+        if (changelogReader != null) {
             changelogReader.unregister(getAllChangelogTopicPartitions());
         }
```
Since `stateUpdaterEnabled` is basically always `true` if we remove the flag, the `if`-condition should always be `false` and the code guarded by the `if`-condition should never be executed.

IMO, the changelog reader can be removed from the `ProcessorStateManager` since registering changelog topics is done in the state updater. Only the old code path that did not use the state updater needed the changelog reader here.

If the `ProcessorStateManager` does not need the changelog reader, the active task creator and the standby task creator do also not need the changelog reader.

cadonna commented on streams/src/main/java/org/apache/kafka/streams/processor/internals/ProcessorStateManager.java:
```diff
@@ -664,7 +646,7 @@ else if (exception instanceof StreamsException)
+        if (changelogReader != null) {
             final List<TopicPartition> allChangelogs = getAllChangelogTopicPartitions();
             changelogReader.unregister(allChangelogs);
         }
```
See my above comment.

cadonna commented on streams/src/main/java/org/apache/kafka/streams/processor/internals/StreamThread.java:
```diff
@@ -372,15 +371,13 @@ public static StreamThread create(final TopologyMetadata topologyMetadata,
         final String threadId = clientId + THREAD_ID_SUBSTRING + threadIdx;
         final String stateUpdaterId = threadId.replace(THREAD_ID_SUBSTRING, STATE_UPDATER_ID_SUBSTRING);
-        final String restorationThreadId = stateUpdaterEnabled ? stateUpdaterId : threadId;
+        final String restorationThreadId = stateUpdaterId;
```
You could directly use `stateUpdaterId` instead of `restorationThreadId` since the distinction between state updater ID and thread ID for restoration does not hold anymore.

cadonna commented on streams/src/main/java/org/apache/kafka/streams/processor/internals/StreamThread.java:
```diff
@@ -1222,15 +1142,15 @@ private long pollPhase() {
+        if (state == State.PARTITIONS_ASSIGNED) {
             // try to fetch some records with zero poll millis
             // to unblock the restoration as soon as possible
             records = pollRequests(Duration.ZERO);
```
You can get rid of this branch. Polling with duration zero during `PARTITIONS_ASSIGNED` only applies to the old code path. With the state updater polling should use the configured poll time as stated on line 1153.

cadonna commented on streams/src/test/java/org/apache/kafka/streams/processor/internals/StoreChangelogReaderTest.java:
```diff
@@ -407,30 +406,6 @@ public long position(final TopicPartition partition) {
-
     @ParameterizedTest
     @EnumSource(value = Task.TaskType.class, names = {"ACTIVE", "STANDBY"})
     public void shouldPollWithRightTimeoutWithStateUpdaterDefault(final Task.TaskType type) {
```
Could you please include `shouldPollWithRightTimeout()` into this test and rename this test to `shouldPollWithRightTimeout()`?

janchilling replied:
@cadonna Not clear. You mean just change the name of `shouldPollWithRightTimeoutWithStateUpdaterDefault()` to `shouldPollWithRightTimeout()` ? or bring back `shouldPollWithRightTimeout()` as it was earlier?

The reason I removed `shouldPollWithRightTimeout()` method was since `shouldPollWithRightTimeoutWithStateUpdaterDefault()` also has the same behavior and they were checking the same thing basically. So redundant code.

cadonna replied:
I meant to copy the content of `shouldPollWithRightTimeout(final Properties properties, final Task.TaskType type)` (below this method) into this method since `shouldPollWithRightTimeout()` is only called in this method. Then rename this method from `shouldPollWithRightTimeoutWithStateUpdaterDefault()` to `shouldPollWithRightTimeout()`.

cadonna commented on streams/src/test/java/org/apache/kafka/streams/processor/internals/StreamThreadTest.java:
```diff
@@ -1716,15 +1699,14 @@ public void shouldReinitializeRevivedTasksInAnyState(final boolean stateUpdaterE
-        if (stateUpdaterEnabled) {
-            TestUtils.waitForCondition(
+        TestUtils.waitForCondition(
                 () -> thread.taskManager().checkStateUpdater(
                     mockTime.milliseconds(),
                     topicPartitions -> mockConsumer.seekToBeginning(singleton(t1p1))
                 ),
                 10 * 1000,
                 "State updater never returned tasks.");
```
Please fix the indentation here.

cadonna commented on streams/src/test/java/org/apache/kafka/streams/processor/internals/StreamThreadTest.java:
```diff
@@ -1736,15 +1718,13 @@ public void shouldReinitializeRevivedTasksInAnyState(final boolean stateUpdaterE
-        if (stateUpdaterEnabled) {
-            TestUtils.waitForCondition(
+        TestUtils.waitForCondition(
                 () -> thread.taskManager().checkStateUpdater(
                     mockTime.milliseconds(),
                     topicPartitions -> mockConsumer.seekToBeginning(singleton(t1p1))
                 ),
                 10 * 1000,
                 "State updater never returned tasks.");
-        }
```
See my comment above about indentation.

cadonna commented on streams/src/test/java/org/apache/kafka/streams/processor/internals/StreamThreadTest.java:
```diff
@@ -2491,26 +2467,17 @@ public Set<TopicPartition> partitions() {
-        if (stateUpdaterEnabled) {
-            TestUtils.waitForCondition(
+        TestUtils.waitForCondition(
                 () -> mockRestoreConsumer.assignment().size() == 0,
                 "Never get the assignment");
```
Fix indentation and could you please also fix the typo? It should be `Never got the assignment`.

cadonna reviewed (commented):
@janchilling Thanks for the updates!

Here my comments.

janchilling committed:
KAFKA-18913: Updates in order to fix the comments raised on the review.

Changes - Indentation fixes in the StoreChangelogReaderTest class
         - Removed restorationThreadId from the StreamThread, instead of it stateUpdaterId used since distinction between state updater ID and thread ID for restoration does not hold anymore.
         - Removed the `if (state == State.PARTITIONS_ASSIGNED && !stateUpdaterEnabled)` from the pollPhase() method in the StreamThread since with state updater polling should use the configured poll time.
         - Merged the shouldPollWithRightTimeout test case with shouldPollWithRightTimeoutWithStateUpdaterDefault to shouldPollWithRightTimeout.

janchilling committed:
KAFKA-18913: Indentation fix

janchilling committed:
KAFKA-18913: Removed the stateUpdaterEnabled flag from all the integration tests.

janchilling committed:
Merge remote-tracking branch 'origin/trunk' into KAFKA-18913/Old-code-removal-gaurded-by-stateUpdaterEnabled-flag

# Conflicts:
#	streams/integration-tests/src/test/java/org/apache/kafka/streams/integration/KafkaStreamsTelemetryIntegrationTest.java

janchilling committed:
KAFKA-18913: Updates to the KafkaStreamsTelemetryIntegrationTest class since they couldn't be passed with merging trunk into this branch due to a merge conflict

This includes a passing null to props method and removing some unused imports.

janchilling committed:
KAFKA-18913: Corrected a file format violation pointed out in the workflow build

janchilling committed:
KAFKA-18913: Corrected the test cases to use the correct parameters and made parameterized tests normal tests if no parameters were required.

janchilling committed:
KAFKA-18913: Removed unused imports pointed out by the github workflow

janchilling committed:
KAFKA-18913: Resolving a issue pointed out by the github workflow

janchilling committed:
KAFKA-18913: Removing changelogReader from ProcessorStateManager, ActiveTaskCreator, StandbyTaskCreator and updates to TaskManager class

Removed the ChangelogReader from the ProcessorStateManager, ActiveTaskCreator, and StandbyTaskCreator and related test cases. This was based on a review in the PR, since the registering with ChangelogReader is done in the old code flow where the state updater is disabled.

Furthermore, the TaskManager class was also updated. And the update includes the removal of code where the condition (stateUpdater == null) is false, since the stateUpdater is not null by default.

janchilling committed:
KAFKA-18913: Removed the needsInitializationOrRestoration() method in the TaskManager since the method is not used anymore as the initializeAndRestorePhase() method was removed from the StreamThread.

janchilling committed:
KAFKA-18913: Removed the STATE_UPDATER_ENABLED config from the StreamsConfig class.

janchilling commented:
Hi @cadonna ,

I have removed the stateUpdaterEnabled flag and the related code completely from the codebase, and only remaining part is to remove it from a readme file and a python test case. And sorry for the force pushes and the PR being a bit messy!

cadonna commented on streams/integration-tests/src/test/java/org/apache/kafka/streams/integration/SmokeTestDriverIntegrationTest.java:
```diff
@@ -99,8 +99,8 @@ SmokeTestDriver.VerificationResult result() {
     @ParameterizedTest
-    @CsvSource({"false, false", "true, false"})
-    public void shouldWorkWithRebalance(final boolean stateUpdaterEnabled, final boolean processingThreadsEnabled) throws InterruptedException {
+    @CsvSource({"true, false"})
```
This does not look right!
It should be
```suggestion
    @CsvSource({"true", "false"})
```
You can see it when you run the test. It should have two runs -- one for processing threads enabled and one for processing threads disabled, but it only has one with processing threads enabled.
With my suggestion, you get the two runs.

cadonna replied:
Now you might be even able to use `@ValueSource(booleans = {true, false})` instead of `@CsvSource`.

cadonna commented on streams/integration-tests/src/test/java/org/apache/kafka/streams/integration/KafkaStreamsTelemetryIntegrationTest.java:
```diff
@@ -432,19 +428,8 @@ private List<String> getTaskIdsAsStrings(final KafkaStreams streams) {
+                Arguments.of("complex"));
     }
 
     private Properties props(final Properties extraProperties) {
```
Since this method is exclusively called with `null` now, you can remove the parameter.

cadonna commented on streams/integration-tests/src/test/java/org/apache/kafka/streams/integration/KafkaStreamsTelemetryIntegrationTest.java:
```diff
@@ -264,9 +261,9 @@ public void shouldPushMetricsToBroker(final String recordingLevel) throws Except
 
     @ParameterizedTest
     @MethodSource("singleAndMultiTaskParameters")
```
This can probably be transformed to `@ValueSource(strings = {"simple", "complex"})`

cadonna commented on streams/integration-tests/src/test/java/org/apache/kafka/streams/integration/RestoreIntegrationTest.java:
```diff
@@ -161,7 +154,10 @@ private Properties props(final Properties extraProperties) {
+
+        if (extraProperties != null) {
+            streamsConfiguration.putAll(extraProperties);
+        }
```
Instead of passing `null` could you please do the following:
```java
    private Properties props(final Properties extraProperties) {
        Properties streamsConfiguration = props();
        streamsConfiguration.putAll(extraProperties);
        return streamsConfiguration;
    }

    private Properties props() {
        final Properties streamsConfiguration = new Properties();

        streamsConfiguration.put(StreamsConfig.APPLICATION_ID_CONFIG, appId);
        streamsConfiguration.put(StreamsConfig.BOOTSTRAP_SERVERS_CONFIG, CLUSTER.bootstrapServers());
        streamsConfiguration.put(StreamsConfig.STATESTORE_CACHE_MAX_BYTES_CONFIG, 0);
        streamsConfiguration.put(StreamsConfig.STATE_DIR_CONFIG, TestUtils.tempDirectory(appId).getPath());
        streamsConfiguration.put(StreamsConfig.DEFAULT_KEY_SERDE_CLASS_CONFIG, Serdes.IntegerSerde.class);
        streamsConfiguration.put(StreamsConfig.DEFAULT_VALUE_SERDE_CLASS_CONFIG, Serdes.IntegerSerde.class);
        streamsConfiguration.put(StreamsConfig.COMMIT_INTERVAL_MS_CONFIG, 1000L);
        streamsConfiguration.put(ConsumerConfig.AUTO_OFFSET_RESET_CONFIG, "earliest");

        streamsConfigurations.add(streamsConfiguration);

        return streamsConfiguration;
    }

```

cadonna commented on streams/src/main/java/org/apache/kafka/streams/processor/internals/TaskManager.java:
```diff
@@ -899,6 +840,7 @@ private StreamTask convertStandbyToActive(final StandbyTask standbyTask, final S
         return activeTaskCreator.createActiveTaskFromStandby(standbyTask, partitions, mainConsumer);
     }
 
+    // This can also be removed, but what about the test cases?
```
If the production code is removed also the test cases are not needed anymore.
If you have comments or questions please add them to the PR instead of inline comments in code. Inline comments are easy to overlook and then we have comments that nobody understands in the codebase.

cadonna commented on streams/src/test/java/org/apache/kafka/streams/processor/internals/StreamThreadTest.java:
```diff
@@ -364,8 +361,8 @@ public void onChange(final Thread thread,
 
     @ParameterizedTest
     @MethodSource("data")
```
You could change this to `@ValueSource(booleans = {true, false})`.

cadonna commented on streams/src/test/java/org/apache/kafka/streams/processor/internals/TaskManagerTest.java:
```diff
@@ -981,7 +933,7 @@ public void shouldCreateStandbyTaskDuringAssignment() {
 
     @Test
-    public void shouldAddRecycledStandbyTasksFromActiveToPendingTasksToInitWithStateUpdaterEnabled() {
+    public void shouldAddRecycledStandbyTasksFromActiveToPendingTasksToInitWithStateUpdater() {
```
```suggestion
    public void shouldAddRecycledStandbyTasksFromActiveToPendingTasksToInit() {
```
Here and elsewhere.

cadonna reviewed (commented):
Thanks for the updates, @janchilling !

I had some comments. I also noticed that there are some `if`-conditions still in the code base like:

```java
if (stateUpdater != null) {
...
```
Could you please also get rid of those?

There are also conflicts that needs to be resolved. Some (if not all) are caused by one of my commits. I am sorry!

cadonna commented:
FYI: I will not be online for the next 1.5 weeks.

janchilling committed:
Merge branch 'trunk' of https://github.com/janchilling/kafka into KAFKA-18913/Old-code-removal-gaurded-by-stateUpdaterEnabled-flag

# Conflicts:
#	streams/integration-tests/src/test/java/org/apache/kafka/streams/integration/SmokeTestDriverIntegrationTest.java
#	streams/src/main/java/org/apache/kafka/streams/processor/internals/StreamThread.java
#	streams/src/test/java/org/apache/kafka/streams/processor/internals/TaskManagerTest.java

janchilling committed:
KAFKA-18913: Addressed the comments in the PR.

janchilling committed:
Merge branch 'trunk' of https://github.com/janchilling/kafka into KAFKA-18913/Old-code-removal-gaurded-by-stateUpdaterEnabled-flag

# Conflicts:
#	streams/src/test/java/org/apache/kafka/streams/processor/internals/StreamThreadTest.java

janchilling committed:
KAFKA-18913: Correct some code error resulting from the merge conflict.

cadonna commented:
@janchilling Is the PR ready for re-review?

janchilling commented:
> @janchilling Is the PR ready for re-review?

Hi @cadonna Not yet, I am having some university exams and assignments to complete, so could not focus on this for the past few weeks. But I'll make sure this is ready by the end of this week.

Have a few questions also, some test cases in the StreamThreadTest is failing due to this line, https://github.com/apache/kafka/blob/trunk/streams/src/test/java/org/apache/kafka/streams/processor/internals/StreamThreadTest.java/#L252 . 
Basically since in some test cases we don't directly create a thread
(https://github.com/apache/kafka/blob/trunk/streams/src/test/java/org/apache/kafka/streams/processor/internals/StreamThreadTest.java/#L3647 ). And I am a bit unsure about what can be done here, so maybe can you give me a pointer?

cadonna commented:
> Hi @cadonna Not yet, I am having some university exams and assignments to complete, so could not focus on this for the past few weeks. But I'll make sure this is ready by the end of this week.

No worries!

janchilling committed:
KAFKA-18913: This commit includes changes to ProcessorStateManagerTest, StreamThreadTest, TaskManagerTest classes and TaskManager class.

ProcessorStateManagerTest class changes :
 - Removed the use of ChangelogReader from the class, since it is not used in the StateUpdater enabled default flow
 - Removed ProcessorStateManagerTest#shouldNotRegisterNonLoggedStore and ProcessorStateManagerTest#shouldUnregisterChangelogsDuringClose, since ChangelogReader was removed.
 - Changed the ProcessorStateManagerTest#shouldRegisterNonPersistentStore and ProcessorStateManagerTest#shouldRegisterPersistentStores test cases to verify the registerStore method.

TaskManagerTest class changes :
 - Removed test cases that verified the old code flow which used the TaskManager#handleAssignment with TaskManager#tryToCompleteRestoration since it follows the old code flow.

TaskManager class changes :
 - Removed TaskManager#tryToCompleteRestoration method since it is a part of the old code flow without the StateUpdater.

StreamThreadTest class changes :
 - Changes @MethodSource("data") to @ValueSource(booleans = {true, false}) since only now true and false only needs to be passed to the test methods since state updater is default true.
 - Removed tests shouldOnlyCompleteShutdownAfterRebalanceNotInProgress, shouldUpdateStandbyTask, shouldNotBlockWhenPollingInPartitionsAssignedStateWithoutStateUpdater and shouldNotUpdateStandbyTaskWhenPaused test cases since they needed the StateUpdater to be false.

 Note - testNamedTopologyWithStreamsProtocol and testStreamsRebalanceDataWithExtraCopartition test will fail since those verify an IllegalStateException when creating a Stream Thread. But during the StreamThread.create() method we create a StateUpdater thread regardless of whether a StreamThread is created or not. Therefore, during the teardown process, a StateUpdater thread is left hanging(This is a bug). Will correct this using a separate PR.

janchilling committed:
KAFKA-18913: Removed an unused import

janchilling committed:
Merge branch 'trunk' of https://github.com/janchilling/kafka into KAFKA-18913/Old-code-removal-gaurded-by-stateUpdaterEnabled-flag

# Conflicts:
#	streams/integration-tests/src/test/java/org/apache/kafka/streams/integration/KafkaStreamsTelemetryIntegrationTest.java
#	streams/src/main/java/org/apache/kafka/streams/processor/internals/TaskManager.java

janchilling committed:
KAFKA-18913: Updated the code after merging after merge conflicts

janchilling commented:
Hi @cadonna ,

I think I have completed all, sorry for the delay though.

Only one problem remains. Which is the `StreamThreadTest#testNamedTopologyWithStreamsProtocol` and `StreamThreadTest#testStreamsRebalanceDataWithExtraCopartition` tests will fail since there will be a `StateUpdater thread` running, which is created through the  `StreamThread#createAndStartStateUpdater`  in the `StreamThread#create`. These tests fails during the teardown process, due to the `StateUpdater thread` being left to run even though an `IllegalStateException` thrown before the `StreamThread` is created from the above 2 test cases. I guess this is a bug since there cannot be a `StateUpdater Thread` running when a `StreamThread` is not created.

I will create a separate PR to the above issue with the solution, I guess then we'll have to merge that PR before merging this.

cadonna commented:
Hi @janchilling ,

Sorry for the long silence but I was quite busy recently.

Yeah, I agree with you about the bug regarding the state updater thread that is not torn down after the `IllegalStateException`. However, rather than surrounding the code in the `StreamThread#create()` with a `try-catch`-clause, I propose to create an `init()` method in the `TaskManager` that starts the state updater. The `init()` method is then called in the beginning of the `run()` method of the stream thread. In such a way we have starting and shutting down the state updater thread encapsulated in the `TaskManager` which ensures that the state updater thread it not started before the stream thread is started.

Could you open a separate PR just for that?
If that solves the issue, we would first merge that PR and then merge this PR on top of the that PR.

cadonna commented on streams/integration-tests/src/test/java/org/apache/kafka/streams/integration/KafkaStreamsTelemetryIntegrationTest.java:
```diff
@@ -446,29 +446,15 @@ private List<String> getTaskIdsAsStrings(final KafkaStreams streams) {
 
     private static Stream<Arguments> singleAndMultiTaskParameters() {
```
This name does not really fit anymore. I propose to rename this method to `topologyComplexityAndRebalanceProtocol`.

cadonna commented on streams/integration-tests/src/test/java/org/apache/kafka/streams/integration/KafkaStreamsTelemetryIntegrationTest.java:
```diff
@@ -484,7 +470,7 @@ private Properties props(final Properties extraProperties) {
         streamsConfiguration.put(StreamsConfig.DEFAULT_CLIENT_SUPPLIER_CONFIG, TestClientSupplier.class);
         streamsConfiguration.put(StreamsConfig.InternalConfig.INTERNAL_CONSUMER_WRAPPER, TestConsumerWrapper.class);
         streamsConfiguration.put(ConsumerConfig.AUTO_OFFSET_RESET_CONFIG, "earliest");
-        streamsConfiguration.putAll(extraProperties);
```
This does not seem right. On line 472 the group protocol config is passed to `props()`, but here it is ignored.

cadonna commented on streams/src/main/java/org/apache/kafka/streams/processor/internals/TaskManager.java:
```diff
@@ -395,18 +379,15 @@ public void handleAssignment(final Map<TaskId, Set<TopicPartition>> activeTasks,
-            failedTasks.putAll(collectExceptionsAndFailedTasksFromStateUpdater());
-        }
+
+        handleTasksWithStateUpdater(
```
Could you please rename this method to `handleTasks()`. We do not need to distinguish the cases with and without state updater.

cadonna commented on streams/src/main/java/org/apache/kafka/streams/processor/internals/TaskManager.java:
```diff
@@ -870,12 +806,8 @@ private Map<TaskId, RuntimeException> closeAndRecycleTasks(final Map<Task, Set<T
-                        tasks.removeTask(oldTask);
-                        tasks.addPendingTasksToInit(Collections.singleton(standbyTask));
-                    } else {
-                        tasks.replaceActiveWithStandby(standbyTask);
```
This [method](https://github.com/apache/kafka/blob/0035ac06d33fbd427605cd107e3a1da1ff2061ce/streams/src/main/java/org/apache/kafka/streams/processor/internals/TasksRegistry.java#L58) from the `TaskRegistry` interface is not used anymore. Could you please remove it and its implementations?

cadonna commented on streams/src/test/java/org/apache/kafka/streams/processor/internals/StreamThreadTest.java:
```diff
@@ -1320,68 +1297,14 @@ public void shouldInjectProducerPerThreadUsingClientSupplierOnCreateIfEosV2Enabl
 
     @ParameterizedTest
-    @MethodSource("data")        
-    public void shouldOnlyCompleteShutdownAfterRebalanceNotInProgress(final boolean stateUpdaterEnabled, final boolean processingThreadsEnabled) throws InterruptedException {
-        // The state updater is disabled for this test because this test relies on the fact the mainConsumer.resume()
-        // is not called. This is not true when the state updater is enabled which leads to
-        // java.lang.IllegalStateException: No current assignment for partition topic1-2.
-        // Since this tests verifies an aspect that is independent from the state updater, it is OK to disable
-        // the state updater and leave the rewriting of the test to later, when the code path for disabled state updater
-        // is removed.
```
I believe, you need to rewrite this test as the comment says. Let me know, if you need some help with that.

cadonna commented on streams/src/test/java/org/apache/kafka/streams/processor/internals/TaskManagerTest.java:
```diff
@@ -1851,29 +1736,7 @@ public void shouldReleaseLockForUnassignedTasksAfterRebalanceWithStateUpdater()
     }
 
     @Test
-    public void shouldReportLatestOffsetAsOffsetSumForRunningTask() throws Exception {
```
Could you please add the following test as a replacement for the this test:
```java
    @Test
    public void shouldComputeOffsetSumForRunningStatefulTask() {
        final StreamTask runningStatefulTask = statefulTask(taskId00, taskId00ChangelogPartitions)
            .inState(State.RUNNING).build();
        final long changelogOffsetOfRunningTask = Task.LATEST_OFFSET;
        when(runningStatefulTask.changelogOffsets())
            .thenReturn(mkMap(mkEntry(t1p0changelog, changelogOffsetOfRunningTask)));
        final TasksRegistry tasks = mock(TasksRegistry.class);
        final TaskManager taskManager = setUpTaskManager(ProcessingMode.AT_LEAST_ONCE, tasks, true);
        when(tasks.allTasksPerId()).thenReturn(mkMap(mkEntry(taskId00, runningStatefulTask)));

        assertThat(
            taskManager.taskOffsetSums(),
            is(mkMap(mkEntry(taskId00, changelogOffsetOfRunningTask)))
        );
    }
```

cadonna commented on streams/src/test/java/org/apache/kafka/streams/processor/internals/TaskManagerTest.java:
```diff
@@ -1909,7 +1772,7 @@ public void shouldComputeOffsetSumForRestoringStandbyTaskWithStateUpdater() thro
 
     @Test
-    public void shouldComputeOffsetSumForRunningStatefulTaskAndRestoringTaskWithStateUpdater() {
+    public void shouldComputeOffsetSumForRunningStatefulTaskAndRestoringTask() {
```
Could you please change 
```java
final long changelogOffsetOfRunningTask = 42L;
``` 
to 
```java
final long changelogOffsetOfRunningTask = Task.LATEST_OFFSET;
```
to make the case more real?

cadonna commented on streams/src/test/java/org/apache/kafka/streams/processor/internals/TaskManagerTest.java:
```diff
@@ -1940,57 +1803,6 @@ public void shouldComputeOffsetSumForRunningStatefulTaskAndRestoringTaskWithStat
     }
 
-    @Test
-    public void shouldSkipUnknownOffsetsWhenComputingOffsetSum() throws Exception {
```
Could you please replace this test with the following:
```java
    @Test
    public void shouldSkipUnknownOffsetsWhenComputingOffsetSum() {
        final StreamTask restoringStatefulTask = statefulTask(taskId01, taskId01ChangelogPartitions)
            .inState(State.RESTORING).build();
        final long changelogOffsetOfRestoringStandbyTask = 84L;
        when(restoringStatefulTask.changelogOffsets())
            .thenReturn(mkMap(
                mkEntry(t1p1changelog, changelogOffsetOfRestoringStandbyTask),
                mkEntry(t1p1changelog2, OffsetCheckpoint.OFFSET_UNKNOWN)
            ));
        final TasksRegistry tasks = mock(TasksRegistry.class);
        final TaskManager taskManager = setUpTaskManager(ProcessingMode.AT_LEAST_ONCE, tasks, true);
        when(stateUpdater.tasks()).thenReturn(Set.of(restoringStatefulTask));

        assertThat(
            taskManager.taskOffsetSums(),
            is(mkMap(
                mkEntry(taskId01, changelogOffsetOfRestoringStandbyTask)
            ))
        );
    }
```
where 
```java
    private final TopicPartition t1p1changelog2 = new TopicPartition("changelog2", 1);
```

cadonna commented on streams/src/test/java/org/apache/kafka/streams/processor/internals/TaskManagerTest.java:
```diff
@@ -2097,105 +1909,9 @@ public void shouldPinOffsetSumToLongMaxValueInCaseOfOverflow() throws Exception
     }
 
-    @Test
-    public void shouldCloseActiveUnassignedSuspendedTasksWhenClosingRevokedTasks() {
```
I believe you cannot just delete all the tests that contain `tryToCompleteRetoration()`. You need to rewrite them. Let me know if you need help.

cadonna reviewed (commented):
Thanks for the updates @janchilling! 
I really appreciate your patience and endurance!

Here my feedback!

Most of the tests in task manager test that you deleted need a rewrite instead of just deleting them.

lucasbru commented:
Hey @janchilling are you still working on this? Just checking.

cadonna commented:
@lucasbru Yes, they are. They were sidetracked by the following PR: https://github.com/apache/kafka/pull/19889#issuecomment-2937205109
See also the comment about availability in the next two weeks.

janchilling committed:
Merge remote-tracking branch 'origin/trunk' into KAFKA-18913/Old-code-removal-gaurded-by-stateUpdaterEnabled-flag

# Conflicts:
#	streams/src/main/java/org/apache/kafka/streams/processor/internals/StreamThread.java
#	streams/src/test/java/org/apache/kafka/streams/processor/internals/StreamThreadTest.java

janchilling committed:
KAFKA-18913: State Updater thread not shutting down during teardown in StreamThreadTest fix

Updated the StreamThread#shouldOnlyCompleteShutdownAfterRebalanceNotInProgress test case and some other changes that were requested.

janchilling commented:
Hi @cadonna ,

I rewrote the `StreamThreadTest#shouldOnlyCompleteShutdownAfterRebalanceNotInProgress `. Can you check if what I have written is correct? I specifically had to check if the **StateUpdater thread** is running, because otherwise the test fail. Therefore I checked if a **StateUpdater thread** is present from the Thread Stacktraces. 

Also, can you help me identify the test cases that needs a rewrite from the `TaskManagerTest ` class (This would be of great help). I will be available from today onwards and will give full focus to this and try to finish this by next week.

janchilling committed:
KAFKA-18913: Updates to address the changes requested from the review and updated to test cases in the ProcessorStateManagerTest class

StreamThreadTest changes : Removed two unused methods addStandbyRecordsToRestoreConsumer() and addActiveRecordsToRestoreConsumer()

TaskManagerTest changes : Added methods shouldComputeOffsetSumForRunningStatefulTask() as a replacement for shouldReportLatestOffsetAsOffsetSumForRunningTask() and shouldSkipUnknownOffsetsWhenComputingOffsetSum() as a replacement for shouldSkipUnknownOffsetsWhenComputingOffsetSum().

TasksRegistry and Tasks changes : Removed unused method replaceActiveWithStandby()

ProcessorStateManagerTest changes : Updated the method shouldUnregisterChangelogsDuringClose(), shouldRecycleStoreAndReregisterChangelog() to shouldCloseStateStoresOnStateManagerClose(), shouldRecycleAndReinitializeStore() respectively since the changelogReader was removed by the ProcessorStateManager.

janchilling commented:
Hi @cadonna ,

I kinda need help with updating the Test cases in the TaskManagerTest class.

cadonna commented:
@janchilling Thank you for your patience! 
I am sorry, I am quite busy at the moment and I am not able to carve out some time to review this PR. \cc @lucasbru @mjsax @bbejeck

mjsax commented:
@janchilling -- Had a very brief look into this PR, and it is very large. I would propose to split it up, into multiple smaller PRs to allow us to make incremental progress.

mjsax commented:
Closing this PR due to inactivity. It did become stale, and somebody else picked up this ticket now.
